### PR TITLE
WIP: Console mode using HTerm.js terminal emulator

### PR DESCRIPTION
**Pull request by fredyshox**
This is an attempt to add console mode support. This would enable to interact with vm using serial console (ex. /dev/ttyS0 on x86_64).

## Current progress

The communication is implemented using named pipes (one for input, one for output), which are managed by `UTMTerminal` objects. These pipes are then configured as character device and passed to qemu using `-serial` option.

The `VMTerminalViewController` integrates [HTerm](https://hterm.org) terminal emulator using WKWebView, and is bridging HTerm and UTMTerminal together.

The guest OS still needs to be configured for this to work - serial console must be enabled on boot ([example on Alpine linux](https://wiki.alpinelinux.org/wiki/Enable_Serial_Console_on_Boot)).

## TODO

- [x] abstract away Spice-related code from `UTMVirtualMachine`, so either console mode of gui mode could be used.
- [x] add keyboard input accessory from `VMDisplayMetalViewController` for special keys. This requires workaround on WKWebView
- [x] show keyboard button, this also requires workaround as `element.focus()` doesn't seem to work on WKWebView
- [x] vm screenshot
- [x] issue with Qemu instance unable to quit 

IMG_6393

### CODE REVIEW

**Comment by fredyshox** [comment]
Ok, It seems to be working well. I’ll mark this draft as completed, as the differences between it and master seems to be growing very rapidly. As for the main changes: Spice specific code is removed from `UTMVirtualMachine`. VM now uses `UTMInputOutput` interface for io services, so `UTMSpiceIO` for full graphic mode (where all spice specific code resides) and `UTMTerminalIO` for console mode. 

Things that does not work or got broken:
- screenshots in console mode
- UTMViewState sync/restoration (this should got out of `UTMVirtualMachine` as it is using spice specific code) 
- serial output restoration when resuming vm

File: Managers/UTMVirtualMachine.m [hunk 1]
```diff
@@ -577,20 +532,20 @@ - (BOOL)savePlist:(NSURL *)path dict:(NSDictionary *)dict withError:(NSError **)
 #pragma mark - View State
 
 - (void)syncViewState {
-    self.viewState.displayOriginX = self.primaryDisplay.viewportOrigin.x;
-    self.viewState.displayOriginY = self.primaryDisplay.viewportOrigin.y;
-    self.viewState.displaySizeWidth = self.primaryDisplay.displaySize.width;
-    self.viewState.displaySizeHeight = self.primaryDisplay.displaySize.height;
-    self.viewState.displayScale = self.primaryDisplay.viewportScale;
+//    self.viewState.displayOriginX = self.primaryDisplay.viewportOrigin.x;
```
Comment (osy86) [inline Managers/UTMVirtualMachine.m:535]: Any reason this chunk of code is disabled?

File: UTM.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1833,7 +1918,7 @@
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = JE2494MFB5;
```
Comment (osy86) [inline UTM.xcodeproj/project.pbxproj:1921]: I wish there's a way for Xcode to not put this into the project file.

Comment (saagarjha) [reply UTM.xcodeproj/project.pbxproj:1921]: xcconfig?

File: Managers/UTMVirtualMachine.m (hunk 1 above)
Comment (osy86) [reply Managers/UTMVirtualMachine.m:535]: I see, the ivars are moved.

File: UTM.xcodeproj/project.pbxproj (hunk 2 above)
Comment (osy86) [reply UTM.xcodeproj/project.pbxproj:1921]: TIL. I will add this in the future.

**Comment by osy86** [comment]
Merged in adfd23fc2ed13bf86e5ed8ba77361405f479b6d6

**Pull request closed without merging**